File: problems/1101-1200/1135_cheapest-spanning-network/solution_prim_heap.js
/**
 * @param {number} n
 * @param {number[][]} links
 * @return {number}
 */
var cheapestSpanningNetwork = function (n, links) {
    // adjacency over n + 1 slots (index 0 unused; nodes are 1-based);
    // each link is filed once per direction
    const adj = Array.from({ length: n + 1 }, () => []);
    for (const [x, y, cost] of links) {
        adj[x].push([cost, y]);
        adj[y].push([cost, x]);
    }

    // Min-heap of [cost, node]
    const heap = [];
    const push = (item) => {
        heap.push(item);
        let i = heap.length - 1;
        while (i > 0) {
            const p = (i - 1) >> 1;
            if (heap[p][0] <= heap[i][0]) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };
    const pop = () => {
        const top = heap[0];
        const last = heap.pop();
        if (heap.length > 0) {
            heap[0] = last;
            let i = 0;
            for (;;) {
                const l = 2 * i + 1;
                const r = l + 1;
                let m = i;
                if (l < heap.length && heap[l][0] < heap[m][0]) m = l;
                if (r < heap.length && heap[r][0] < heap[m][0]) m = r;
                if (m === i) break;
                [heap[m], heap[i]] = [heap[i], heap[m]];
                i = m;
            }
        }
        return top;
    };

    const visited = new Array(n + 1).fill(false);
    let total = 0;
    let settled = 0;
    // Prim: grow one tree outward from node 1; the cheapest offer
    // leaving the tree is always safe to buy
    push([0, 1]);
    while (heap.length > 0 && settled < n) {
        const [cost, v] = pop();
        // stale-entry guard: v already joined via an offer at most this cheap
        if (visited[v]) {
            continue;
        }
        visited[v] = true;
        total += cost;
        settled++;
        for (const [w, u] of adj[v]) {
            if (!visited[u]) {
                push([w, u]);
            }
        }
    }
    // queue drained before every node joined: the catalogue cannot
    // connect all n nodes
    return settled === n ? total : -1;
};

File: problems-originals/1101-1200/1135_connecting-cities-with-minimum-cost/solution_prim_heap.cpp
class Solution {
  public:
    int minimumCost(int n, vector<vector<int>> &connections) {
        // adjacency over n + 1 slots (index 0 unused; nodes are 1-based);
        // each link is filed once per direction
        vector<vector<pair<int, int>>> adj(n + 1);
        for (const auto &link : connections) {
            adj[link[0]].push_back({link[2], link[1]});
            adj[link[1]].push_back({link[2], link[0]});
        }
        // Prim: grow one tree outward from node 1; the cheapest offer
        // leaving the tree is always safe to buy
        priority_queue<tuple<int, int>, vector<tuple<int, int>>, greater<tuple<int, int>>> heap;
        heap.emplace(0, 1);
        vector<bool> visited(n + 1, false);
        int total = 0;
        int settled = 0;
        while (!heap.empty() && settled < n) {
            auto [cost, v] = heap.top();
            heap.pop();
            // stale-entry guard: v already joined via an offer at most
            // this cheap
            if (visited[v]) {
                continue;
            }
            visited[v] = true;
            total += cost;
            settled++;
            for (const auto &[w, u] : adj[v]) {
                if (!visited[u]) {
                    heap.emplace(w, u);
                }
            }
        }
        // queue drained before every node joined: the catalogue cannot
        // connect all n nodes
        return settled == n ? total : -1;
    }
};
